---
title: "202503_TCRseq_reanalysis"
output: html_document
date: "2025-03-27"
---

```{r}
library(Seurat)
library(dplyr)
library(purrr)
library(data.table)
library(ggplot2)
library(ggrastr)
library(ggpubr)
library(pals)
library(scCustomize)
#library(data.table)
library(RColorBrewer)
```

```{r, Surat process function}
SeuratProc_T <- function(sobj, verbose=FALSE, dims.clustering=NULL, resolution.clustering=NULL, dims.umap=NULL) {
  sobj <- DietSeurat(sobj)
  DefaultAssay(sobj) <- "RNA"
  sobj <- FindVariableFeatures(sobj, selection.method = "vst", nfeatures = 2000, verbose=verbose)
  sobj <- ScaleData(sobj, features = rownames(sobj), verbose=verbose)
  sobj <- CellCycleScoring(sobj, s.features = cc.genes$s.genes, g2m.features = cc.genes$g2m.genes, set.ident = TRUE)
  sobj <- ScaleData(sobj, vars.to.regress = c("S.Score", "G2M.Score", "percent.mt"), verbose=verbose)
  sobj <- RunPCA(sobj, features = VariableFeatures(sobj), reduction.name = "pcaRNA", reduction.key = "pcaRNA_")
  sobj <- FindNeighbors(sobj, dims = dims.clustering, verbose=verbose, reduction = "pcaRNA")
  sobj <- FindClusters(sobj, resolution = resolution.clustering, verbose=verbose)
  sobj <- RunUMAP(sobj, dims = dims.umap, verbose=verbose, reduction.key = "umapRNA_", reduction.name = "umapRNA", reduction = "pcaRNA")
  return(sobj)
}
```

```{r}
sample_list <- c("LN0132", "LN0144", "LN0178", "LN0193", "LN0198", 
                 "LN0243", "LN0249", "LN0259", "LN0264", "LN0278", "LN0417")

dir_path <- "/omics/groups/OE0606/internal/amathiou/Projects/202402_PhDpapers/LNStroma/202503_TCRseq/"
contig_path <- file.path(dir_path, "contig_files")

Combined_T <- readRDS(file.path(dir_path, "20220901_cd8_citeseq_subclustering.rds"))

sobjs_T <- lapply(sample_list, 
                  function(sample_id) {
  print(paste0('Reading in sample: ', sample_id))
  file <- file.path(dir_path, paste0('SeuratObjectProc_', sample_id, ".rds"))
  sobj <- readRDS(file)
  sobj$PatientID <- sample_id
  sobj[["percent.mt"]] <- PercentageFeatureSet(sobj, pattern = "^MT-")
  sobj <- NormalizeData(sobj)
  sobj <- SeuratProc_T(sobj, 
                       verbose=FALSE, 
                       dims.clustering=1:14, 
                       resolution.clustering=0.4, 
                       dims.umap=1:13)
  return(sobj)
})
names(sobjs_T) <- sample_list
#saveRDS(sobjs_T, file.path(dir_path, "sobjs_Tmapped.rds"))
```

```{r, merge the list of objects}
dir_path <- "/omics/groups/OE0606/internal/amathiou/Projects/202402_PhDpapers/LNStroma/202503_TCRseq/"
sobjs_T <- readRDS(file.path(dir_path, "sobjs_Tmapped.rds"))

sobj <- Merge_Seurat_List(sobjs_T[sample_list])
rm(Combined_T)
rm(sobjs_T)
DefaultAssay(sobj) <- "RNA"
sobj <- FindVariableFeatures(sobj, selection.method = "vst", nfeatures = 2000, verbose=T)

sobj <- ScaleData(sobj, verbose=T)
# Assess cell cycle
sobj <- CellCycleScoring(sobj, s.features = cc.genes$s.genes, g2m.features = cc.genes$g2m.genes, set.ident = TRUE)
sobj <- ScaleData(sobj, vars.to.regress = c("S.Score", "G2M.Score", "percent.mt"), verbose=T)
# Run PCA
sobj <- RunPCA(sobj, features = VariableFeatures(sobj), nfeatures.print=5, ndims.print=1:2,
               reduction.name = "pcaRNA", reduction.key = "pcaRNA_")
# Run clustering based on transcriptome
sobj <- FindNeighbors(sobj, verbose=T, reduction = "pcaRNA")
# Run UMAP based on transcriptome
sobj <- RunUMAP(sobj, dims = 1:13, 
                verbose=T, 
                reduction.key = "umapRNA_",
                reduction.name = "umapRNA", 
                reduction = "pcaRNA")
#saveRDS(sobj, file = '/omics/groups/OE0606/internal/amathiou/Projects/202402_PhDpapers/LNStroma/202503_TCRseq/sobs_scaled.rds')
```


```{r}
sobj <- readRDS('/omics/groups/OE0606/internal/amathiou/Projects/202402_PhDpapers/LNStroma/202503_TCRseq/sobs_scaled.rds')
cd8em <- readRDS(file.path(dir_path, "20220901_cd8_citeseq_subclustering.rds"))#this was combined_T
```


```{r}
all_genes <- rownames(cd8em)
noise_genes <- all_genes[c(grep("^RPS", all_genes), grep("^RPL", all_genes), grep("^MT-", all_genes), 
                           grep("^AC0", all_genes), grep("^AL0", all_genes), grep("^AC1", all_genes), 
                           grep("^AP0", all_genes), grep("^LINC", all_genes), grep("^AL([0-9]+)", all_genes), 
                           grep("^BX([0-9]+)", all_genes), grep("^C([0-9]+)", all_genes), grep("orf", all_genes))]
DefaultAssay(cd8em) <- "RNA"
cd8em <- SCTransform(cd8em, vars.to.regress = "percent.mt", verbose = T)
DefaultAssay(cd8em) <- "integratedRNA"
cd8em <- FindVariableFeatures(cd8em)
cd8em <- RunPCA(object = cd8em, verbose = FALSE, features = setdiff(VariableFeatures(object = cd8em), noise_genes))
cd8em <- RunUMAP(object = cd8em, reduction = "pca", dims = 1:50, return.model=TRUE) #, spread = 1, min.dist = 0.0000001, a = 0.12, b = 0.9
```

```{r}
cd8em@meta.data$CellType <- Idents(cd8em)
DefaultAssay(cd8em) <- "integratedRNA"
DefaultAssay(sobj) <- "RNA"
anchors <- FindTransferAnchors(reference = cd8em,
                               query = sobj, 
                               reference.reduction = "pcaRNA")
sobj <- MapQuery(anchorset = anchors, 
                 reference = cd8em, 
                 query = sobj, 
                 reduction.model = "umap",
                 refdata = list(celltype = "CellType"),
                 reference.reduction = "pcaRNA")
```

```{r}
saveRDS(sobj, file = '/omics/groups/OE0606/internal/amathiou/Projects/202402_PhDpapers/LNStroma/202503_TCRseq/sobs_scaled.rds')
saveRDS(cd8em, file = '/omics/groups/OE0606/internal/amathiou/Projects/202402_PhDpapers/LNStroma/202503_TCRseq/cd8em_formapping.rds')
```

```{r, Read in objects}
sobj <- readRDS("/omics/groups/OE0606/internal/amathiou/Projects/202402_PhDpapers/LNStroma/202503_TCRseq/sobs_scaled.rds")
```

```{r, Read in objects}
cd8em <- readRDS("/omics/groups/OE0606/internal/amathiou/Projects/202402_PhDpapers/LNStroma/202503_TCRseq/cd8em_formapping.rds")
```

```{r, Define DF_TCRrep}
#define the DF_TCRep part
dir_path <- "/omics/groups/OE0606/internal/amathiou/Projects/202402_PhDpapers/LNStroma/202503_TCRseq/"
contig_path <- file.path(dir_path, "contig_files")

readTCR <- function(files = NULL) {
  if (is.null(files) || length(files) == 0) {
    stop("No input files provided to readTCR()")
  }

  lapply(files, fread) %>%
    bind_rows() %>%
    as_tibble() %>%
    mutate(
      PatientID = sapply(strsplit(barcode, "_"), `[`, 2)
    ) #%>%
    #rename(Barcode_full = barcode)
}

tcr_files <- list.files(path = contig_path, pattern = "^filtered_contig_annotations_new_.*\\.csv$", full.names = TRUE) #From Roider et al 2024
#subset for those that are here
sample_list <- c("LN0132", "LN0144", "LN0178", "LN0193", "LN0198", 
                 "LN0243", "LN0249", "LN0259", "LN0264", "LN0278", 
                 "LN0417")
tcr_files <- tcr_files[grepl(paste(sample_list, collapse = "|"), tcr_files)]
```

```{r, Define DF_TCRrep}
DF_TCRrep <- readTCR(tcr_files)
DF_TCRrep$Barcode_full <- DF_TCRrep$barcode
DF_TCRrep#105,706

##### !!!!! ######
#Subset the DF_TCRrep cause the contig folder has more files
##### Keep only cells in ur object
DF_TCRrep <- DF_TCRrep[which(DF_TCRrep$Barcode_full %in% sub("_", "", rownames(sobj@meta.data))), ]
#DF_TCRrep #101,215
```

```{r}
#Patient annotation
entity_map <- tibble::tibble(
  PatientID = c("LN0132", "LN0144", "LN0178", "LN0193", "LN0198", 
                "LN0243", "LN0249", "LN0259", "LN0264", "LN0278", "LN0417"),
  Entity = c("rLN", "FL", "DLBCL", "DLBCL", "FL",
             "FL", "FL", "rLN", "rLN", "FL", "DLBCL")
)

merged_meta <- data.frame(
  Barcode_full = rownames(sobj@meta.data),
  PatientID = sobj@meta.data$PatientID,  # or sobj$sample, etc.
  predicted.celltype = sobj$predicted.celltype,
  stringsAsFactors = FALSE
)
```

```{r}
umap <- sobj@reductions$ref.umap@cell.embeddings
umap_df <- as.data.frame(umap)
umap_df$Barcode_full <- rownames(umap_df)
combined_meta <- merge(merged_meta, umap_df, by = "Barcode_full")
combined_meta$Barcode_full <- sub("_", "", combined_meta$Barcode_full)
```

```{r}
# Merge with your TCR data
DF_TCRrep <- merge(DF_TCRrep, 
                   combined_meta, 
                   by = "Barcode_full")
```

```{r}
DF_TCRrep$PatientID.x <- NULL
DF_TCRrep$PatientID <- DF_TCRrep$PatientID.y
DF_TCRrep$PatientID.y <- NULL
DF_TCRrep <- DF_TCRrep %>% 
  left_join(entity_map, by = "PatientID")
```

```{r}
DF_TCRrep$Entity <- gsub("DLBCL, GCB", "DLBCL", DF_TCRrep$Entity)
sobj@meta.data$Entity <- gsub("DLBCL, GCB", "DLBCL", sobj@meta.data$Entity)
```

```{r}
#Subset it to have only the EM cells
#DF_TCRrep <- DF_TCRrep[!DF_TCRrep$predicted.celltype %in% "T-CD8_Naive", ]
```

```{r}
df_clonotypes <- DF_TCRrep %>% 
  #filter(!is.na(raw_clonotype_id)) %>% 
  filter(!is.na(Entity)) %>% 
  select(barcode, PatientID, raw_clonotype_id, refUMAP_1, refUMAP_2, predicted.celltype, Entity) %>% 
  distinct() %>% 
  add_count(PatientID, raw_clonotype_id, predicted.celltype) %>% 
  group_by(PatientID, raw_clonotype_id, predicted.celltype) %>% 
  summarise(refUMAP_1 = median(refUMAP_1),
            refUMAP_2 = median(refUMAP_2),
            n = first(n),
            Entity = first(Entity),
            .groups = "drop") %>% 
  mutate(n = ifelse(n > 50, 50, n))#clonotypes size maxes out at 50
```

###### UMAPs on EXPRESSION #####
```{r}
#cd8em <- readRDS(file.path(dir_path, "20220901_cd8_citeseq_subclustering.rds"))#this was combined_T
all_genes <- rownames(cd8em)
noise_genes <- all_genes[c(grep("^RPS", all_genes), grep("^RPL", all_genes), grep("^MT-", all_genes), 
                           grep("^AC0", all_genes), grep("^AL0", all_genes), grep("^AC1", all_genes), 
                           grep("^AP0", all_genes), grep("^LINC", all_genes), grep("^AL([0-9]+)", all_genes), 
                           grep("^BX([0-9]+)", all_genes), grep("^C([0-9]+)", all_genes), grep("orf", all_genes))]
DefaultAssay(cd8em) <- "RNA"
cd8em <- SCTransform(cd8em, vars.to.regress = "percent.mt", verbose = T)
DefaultAssay(cd8em) <- "integratedRNA"
cd8em <- FindVariableFeatures(cd8em)
cd8em <- RunPCA(object = cd8em, verbose = FALSE, features = setdiff(VariableFeatures(object = cd8em), noise_genes))
cd8em <- RunUMAP(object = cd8em, reduction = "pca", dims = 1:50, return.model=TRUE) #, spread = 1, min.dist = 0.0000001, a = 0.12, b = 0.9

DefaultAssay(cd8em) <- "RNA"
cd8em <- AddMetaData(cd8em, metadata = FetchData(cd8em, vars = c("CXCL13")))

cd8em <- AddMetaData(
  cd8em,
  metadata = Embeddings(cd8em, reduction = "umap")[, 1:2]
)
```

#UMAP Figure 2 on citeseq data (CXCL13)
```{r}
pdf("/omics/groups/OE0606/internal/amathiou/Projects/202402_PhDpapers/LNStroma/202503_TCRseq/CXCL13_UMAP_CITEseqobj.pdf", width = 4.5, height = 3)
ggplot(cd8em@meta.data %>%
  arrange(CXCL13 >= 2), 
       aes(x = umap_1, y = umap_2, color = CXCL13)) +
  geom_point(size = 0.5, stroke=0) +
  scale_color_gradientn(colors = c("#ECECEC", "#AD2E24")) + 
  theme_classic()
dev.off()
```

#UMAP Figure 3 on TCR data (IFNG)
```{r}
DefaultAssay(sobj) <- "RNA"
sobj <- AddMetaData(sobj, metadata = FetchData(sobj, vars = c("IFNG")))

sobj <- AddMetaData(
  sobj,
  metadata = Embeddings(sobj, reduction = "ref.umap")[, 1:2]
)
```

```{r}
#pdf("/omics/groups/OE0606/internal/amathiou/Projects/202402_PhDpapers/LNStroma/202503_TCRseq/ifngumap_tcrobj_split.pdf", width = 9, height = 3)
ggplot(sobj@meta.data, 
       aes(x = refUMAP_1, y = refUMAP_2, color = IFNG)) +
  geom_point(size = 0.5, stroke=0) +
  scale_color_gradientn(colors = c("#ECECEC", "#AD2E24")) + 
  theme_classic() + 
  facet_wrap(~Entity)
#dev.off()
```

```{r}
#pdf("/omics/groups/OE0606/internal/amathiou/Projects/202402_PhDpapers/LNStroma/202503_TCRseq/ifngumap_tcrobj.pdf", width = 9, height = 3)
ggplot(sobj@meta.data, 
       aes(x = refUMAP_1, y = refUMAP_2, color = IFNG)) +
  geom_point(size = 0.5, stroke=0) +
  scale_color_gradientn(colors = c("#ECECEC", "#AD2E24")) + 
  theme_classic()
#dev.off()
```

#UMAP Figure 3 on TCR data - clone size
```{r, Figure 3 UMAP}
all_barcodes <- sub("_", "", rownames(sobj@meta.data))
tcrep_barcodes <- unique(DF_TCRrep$Barcode_full)
missing_barcodes <- setdiff(all_barcodes, tcrep_barcodes)

# Extract metadata for missing cells since not all cells have a clonotype 
missing_meta <- sobj@meta.data[which(sub("_", "", rownames(sobj@meta.data)) %in% missing_barcodes), ]

########## Add refUMAP coordinates
umap_coords <- sobj@reductions$ref.umap@cell.embeddings
missing_umap <- umap_coords[rownames(missing_meta), ]

# Create combined dataframe for missing cells ---
missing_df <- data.frame(
  refUMAP_1 = missing_umap[, 1],
  refUMAP_2 = missing_umap[, 2],
  predicted.celltype = missing_meta$predicted.celltype,  # adjust if column name differs
  Entity = factor(missing_meta$Entity, levels=c("rLN", "FL", "DLBCL", "DLBCL, GCB")),
  PatientID = missing_meta$PatientID,
  row.names = NULL
)

# Create your original plotting dataframe
DF_5prime_umap <- DF_TCRrep %>% 
  filter(!is.na(Entity)) %>%
  dplyr::select(refUMAP_1, refUMAP_2, predicted.celltype, Entity, PatientID) %>% 
  distinct() %>% 
  mutate(Entity=factor(Entity, levels=c("rLN", "FL", "DLBCL", "DLBCL, GCB")))

# Combine both
DF_5prime_umap_cd8_full <- bind_rows(DF_5prime_umap, missing_df)
```


```{r, Figure 3 UMAP}
p1 <- ggplot() +
  geom_point(data=DF_5prime_umap_cd8_full, aes(x=refUMAP_1, y=refUMAP_2), 
             size=1, alpha=1, color = '#ECECEC', stroke = 0) +
  geom_point(data=df_clonotypes,
             aes(x=refUMAP_1, y=refUMAP_2, size=n, color=Entity), shape=21, stroke=0.7, alpha=1) +
  scale_color_manual(values = c("FL" = "#8BCDE0", "rLN" = "#8D60B5", "DLBCL" = "#67B86E", 'DLBCL, GCB'="#67B86E"), guide="none")+
  scale_fill_manual(values = c("FL" = "#8BCDE0", "rLN" = "#8D60B5", "DLBCL" = "#67B86E", 'DLBCL, GCB'="#67B86E"), guide="none")+
  scale_size_continuous(range=c(1, 10), 
                        limits=c(3, 50), 
                        breaks=c(3, 20, 35, 50),
                        labels=c("3", "20", "35", "> 50"), 
                        name = NULL)+
  #facet_wrap(~Entity, nrow = 1)+
  geom_text(data = df_clonotypes %>% dplyr::select(PatientID, Entity) %>% 
              distinct() %>% 
              add_count(Entity), aes(label = paste0("n = ", n)),
            x = 9.65, y = -6.9, hjust=1, check_overlap = T, size=2.25)+
  xlab("refUMAP-1")+
  ylab("refUMAP-2")+
  theme_classic() +
  theme(legend.position = "right")
```


```{r, Figure 3 UMAP}
pdf("/omics/groups/OE0606/internal/amathiou/Projects/202402_PhDpapers/LNStroma/202503_TCRseq/202310_TCRseq_UMAP_allCD8.pdf", width = 5, height = 3)
plot(p1)
dev.off()
```

```{r}
# Normalize DF_TCRrep barcodes (only)
DF_TCRrep$Barcode_full <- sub("^_", "", DF_TCRrep$Barcode_full)
# Create normalized copy of sobj barcodes for comparison
sobj_barcodes <- sub("^_", "", rownames(sobj@meta.data))
# Find missing barcodes
missing_barcodes <- setdiff(sobj_barcodes, DF_TCRrep$Barcode_full)
# Get original barcodes from sobj that match the missing normalized ones
original_missing_barcodes <- rownames(sobj@meta.data)[sub("^_", "", rownames(sobj@meta.data)) %in% missing_barcodes]
missing_meta <- sobj@meta.data[original_missing_barcodes, ]

# Add metadata
umap_coords <- sobj@reductions$ref.umap@cell.embeddings
missing_umap <- umap_coords[rownames(missing_meta), ]

missing_df <- data.frame(
  Barcode_full = sub("^_", "", original_missing_barcodes),
  Entity = missing_meta$Entity,
  predicted.celltype = missing_meta$predicted.celltype,
  PatientID = missing_meta$PatientID, 
  refUMAP_1 = missing_umap[, 1],
  refUMAP_2 = missing_umap[, 2]
)

# Add placeholder TCR columns as NA
missing_df$raw_clonotype_id <- NA

# Get all column names in DF_TCRrep
all_cols <- colnames(DF_TCRrep)

# Add any missing columns to missing_df and fill with NA
for (col in setdiff(all_cols, colnames(missing_df))) {
  missing_df[[col]] <- NA
}

# Reorder columns to match DF_TCRrep
missing_df <- missing_df[, all_cols]

# Combine safely
DF_TCRrep_cd8 <- bind_rows(DF_TCRrep, missing_df)
```

#From now on work with DF_TCRrep_cd8

```{r}
clonotype_df <- as.data.frame.matrix(table(DF_TCRrep_cd8$raw_clonotype_id, DF_TCRrep_cd8$Entity))
clonotype_freq <- as.data.frame(rowSums(clonotype_df))
colnames(clonotype_freq) <- "Freq"
```


```{r}
pdf("/omics/groups/OE0606/internal/amathiou/Projects/202402_PhDpapers/LNStroma/202503_TCRseq/SFig10_panelA_clonesize_newER_threshold6_woNaive.pdf", width = 6, height = 4)
ggplot(clonotype_freq, aes(x = log10(Freq) )) + 
  geom_density() + 
  theme_bw() + 
  geom_vline(xintercept = log10(6), 
             linetype="dashed", color = "#7F0000", size=1) +
  xlab("log10(Freq_Clone)") + 
  ggtitle("6,349 clonotypes")
dev.off()
```


```{r}
#From seurat object extract cxcl13 and ifng expression
merged_exp <- FetchData(sobj, vars = c('rna_IFNG', 'CXCL13'))
colnames(merged_exp) <- c('IFNG', 'CXCL13')
```

```{r}
merged_exp$Barcode_full <- rownames(merged_exp)
merged_exp$Barcode_full <- sub("^_", "", merged_exp$Barcode_full)
DF_TCRrep_cd8 <- merge(DF_TCRrep_cd8, merged_exp, by = "Barcode_full")
```

```{r}
DF_TCRrep_cd8$IFNG.x <- NULL
DF_TCRrep_cd8$IFNG <- DF_TCRrep_cd8$IFNG.y
DF_TCRrep_cd8$IFNG.y <- NULL
DF_TCRrep_cd8$CXCL13.x <- NULL
DF_TCRrep_cd8$CXCL13 <- DF_TCRrep_cd8$CXCL13.y
DF_TCRrep_cd8$CXCL13.y <- NULL
```

```{r}
df_clonotypes <- 
  DF_TCRrep_cd8 %>% 
  #filter(!is.na(raw_clonotype_id)) %>% 
  filter(!is.na(Entity)) %>% 
  mutate(Entity=factor(Entity, levels=c("rLN", 
                                        "FL", 
                                        "DLBCL")))
```

```{r}
clone_df <- data.frame(Clone = unique(df_clonotypes$raw_clonotype_id))

clone_df$Freq <- unlist(lapply(unique(df_clonotypes$raw_clonotype_id), 
                               function(clonotype){
  nrow(df_clonotypes[which(df_clonotypes$raw_clonotype_id == clonotype), ])
  }))

clone_df$Expanded <- ifelse(clone_df$Freq > 6, 
                            "Expanded", 
                            "Non-expanded")
```

```{r}
clone_df$IFNG <- unlist(lapply(unique(df_clonotypes$raw_clonotype_id), function(clonotype){
  mean(df_clonotypes[which(df_clonotypes$raw_clonotype_id == clonotype), "IFNG"])
  }))
```

```{r}
pdf("/omics/groups/OE0606/internal/amathiou/Projects/202402_PhDpapers/LNStroma/202503_TCRseq/SFig10_panelB_clonesize_ifng_correlation_threshold6_woNaive.pdf", width = 6, height = 4)

ggplot(clone_df, aes(x = log10(Freq), y = IFNG, color = Expanded, fill = Expanded)) + 
  geom_point() + 
  theme_bw() + 
  ggtitle("CD8+ T cells") + 
  xlab("log10(Clone Size)") + 
  geom_smooth(method = "lm", se = TRUE, fullrange = F) + 
  scale_color_manual(values = c("#7F0000", "darkgrey")) + 
  scale_fill_manual(values = c("#7F0000", "darkgrey"))
dev.off()
```

```{r}
clone_df <- data.frame(Clone = unique(df_clonotypes$raw_clonotype_id))

# Mean IFNG expression per clone
clone_df$IFNG <- unlist(lapply(clone_df$Clone, 
                               function(clonotype) {
  mean(df_clonotypes[df_clonotypes$raw_clonotype_id == clonotype, "IFNG"])
}))

# Clone size (frequency)
clone_df$Freq <- unlist(lapply(clone_df$Clone, function(clonotype) {
  nrow(df_clonotypes[df_clonotypes$raw_clonotype_id == clonotype, ])
}))

# Expansion status
clone_df$Expanded <- ifelse(clone_df$Freq > 6, "Expanded", "Non-expanded")
```


```{r}
clones_meta <- clone_df[, c("Clone", "Expanded")]
colnames(clones_meta) <- c("raw_clonotype_id", "Expanded")

df_clonotypes <- merge(df_clonotypes, 
                       clones_meta, by = "raw_clonotype_id")

clone_entity_bar <- as.data.frame(table(df_clonotypes$Expanded, df_clonotypes$Entity))
colnames(clone_entity_bar) <- c("Expanded", "Entity", "Freq")
```

```{r}
df_clonotypes <- DF_TCRrep_cd8 %>%
  filter(!is.na(raw_clonotype_id)) %>%
  filter(!is.na(Entity)) %>%
  mutate(Entity = ifelse(Entity == "DLBCL, GCB", "DLBCL", Entity))

clone_df <- df_clonotypes %>%
  group_by(raw_clonotype_id) %>%
  summarise(
    Freq = n(),
    IFNG = mean(IFNG, na.rm = TRUE)
  ) %>%
  mutate(
    Expanded = ifelse(Freq > 6, "Expanded", "Non-expanded")
  )


# Map back to each cell
expanded_df <- DF_TCRrep[, c("Barcode_full", "raw_clonotype_id")]
expanded_df <- merge(expanded_df, 
                     clone_df[, c("raw_clonotype_id", "Expanded")], 
                     by = "raw_clonotype_id", 
                     all.x = TRUE)

# Mark cells with no clonotype as Non-expanded
expanded_df$Expanded[is.na(expanded_df$Expanded)] <- "Non-expanded"


# Fetch IFNG + meta from sobj
ifng_df <- FetchData(sobj, vars = c("IFNG"))
ifng_df$Barcode_full <- sub("^_", "", rownames(ifng_df))
ifng_df$Entity <- sobj@meta.data[rownames(ifng_df), "Entity"]
ifng_df$predicted.celltype <- sobj@meta.data[rownames(ifng_df), "predicted.celltype"]
ifng_df$Entity <- ifelse(ifng_df$Entity == "DLBCL, GCB", "DLBCL", ifng_df$Entity)  # Normalize


# Merge IFNG/meta with expanded info
plot_df <- merge(ifng_df, expanded_df[, c("Barcode_full", "Expanded")],
                 by = "Barcode_full", all.x = TRUE)

# Assign Non-expanded to any missing
plot_df$Expanded[is.na(plot_df$Expanded)] <- "Non-expanded"

# Set factor levels
plot_df$Expanded <- factor(plot_df$Expanded, levels = c("Non-expanded", "Expanded"))
plot_df$Entity <- factor(plot_df$Entity, levels = c("rLN", "FL", "DLBCL"))

#I think barcode full can be duplicated in there 
plot_df <- distinct(plot_df)
```


```{r}
pdf("/omics/groups/OE0606/internal/amathiou/Projects/202402_PhDpapers/LNStroma/202503_TCRseq/Figure3_BoxPlot_IFNG_threshold6_woNaive.pdf", width = 10, height = 4)
ggplot(plot_df, aes(x = Expanded, y = IFNG)) + 
  geom_boxplot(aes(color = Expanded, fill = Expanded), 
               alpha = 1, outlier.size = 0.1, size = 0.5) + 
  theme_classic() + 
  facet_wrap(~Entity) + 
  scale_color_manual(values = c("black", "black")) + 
  scale_fill_manual(values = c("Non-expanded" = "lightgrey", 
                               "Expanded" = "#7F0000")) + 
  stat_compare_means()
dev.off()
```

#Figure 3
```{r}
ggplot(plot_df, aes(x = Expanded, y = IFNG)) + 
  geom_boxplot(aes(color = Expanded, fill = Expanded), 
               alpha = 1, outlier.size = 0.1, size = 0.5) + 
  theme_classic() + 
  facet_wrap(~Entity) + 
  scale_color_manual(values = c("black", "black")) + 
  scale_fill_manual(values = c("Non-expanded" = "lightgrey", 
                               "Expanded" = "#7F0000")) + 
  stat_compare_means()
```

```{r}
clone_entity_bar <- plot_df %>%
  group_by(Entity, Expanded) %>%
  summarise(Freq = n(), .groups = "drop")

clone_entity_bar <- clone_entity_bar %>%
  mutate(
    Entity = factor(Entity, levels = c("rLN", "FL", "DLBCL")),
    Expanded = factor(Expanded, levels = c("Non-expanded", "Expanded"))
  )
```


```{r}
pdf("/omics/groups/OE0606/internal/amathiou/Projects/202402_PhDpapers/LNStroma/202503_TCRseq/Figure3_Barplot_threshold6_woNaive.pdf", width = 6, height = 4)
ggplot(clone_entity_bar, 
       aes(x = Entity, y = Freq, fill = Expanded, color = Expanded, label = Freq)) + 
  geom_bar(stat = "identity", position = "fill", alpha = 1, size = 0.5) + 
  scale_fill_manual(values = c("Non-expanded" = "lightgrey", "Expanded" = "#7F0000")) + 
  scale_color_manual(values = c("Non-expanded" = "black", "Expanded" = "black")) + 
  theme_classic() + 
  ylab("Proportion of Cells") +
  xlab("Entity") +
  guides(color = "none")  # optional: hides color border legend
dev.off()
```
