---
title: "Figure2"
output: html_document
date: "2024-02-29"
---

# Load pacakges
```{r setup, include=FALSE}
library(Seurat)
library(data.table)
library(ggplot2)
library(tidyverse)
library(dplyr)
library(readr)
library(tidyverse)
library(org.Hs.eg.db)
library(GO.db)
library(ggpubr)
library(patchwork)
```

# Load objects
```{r setup, include=FALSE}
scRNA <- readRDS("") #Available in zenodo as scRNA_LN_stromal_cells.rds
out_dir <- "/g/scb2/zaugg/amathiou/2021July_LNS012/" #Needs to be defined
source("/path/to/Figure_config.R")
```

# Panels 
## A
```{r  Panel A, fig.height=8, fig.width=8, fig.cap="UMAP of LN stronal cells of RNA sequencing data with differential abundand cells"}
## Subset to stromal cells
stroma_obj <- subset(
  scRNA,
  subset = Abe_binned %in% c("BEC", "LEC", "Mesenchymal")
)

custom_colors <- c(
  "DLBCL" = "#67b86e",
  "FL"    = "#8bcde0",
  "rLN"   = "#8d60b5"
)

# Left 
DimPlot(stroma_obj, group.by='group', cols=custom_colors)

# Right 
## Build frequency table 
freq_df <- as.data.frame(table(
  stroma_obj@meta.data$Clusters_paper,
  stroma_obj@meta.data$group,
  stroma_obj@meta.data$patient
))

colnames(freq_df) <- c("Cluster", "Group", "Patient", "Freq")
#LECs not plotted
freq_df <- freq_df %>%
  dplyr::filter(Cluster != "LEC")

wide_df <- freq_df %>%
  dplyr::filter(Cluster %in% c("BEC", "rBEC", "FRC", "rFRC", "FDC")) %>%
  tidyr::pivot_wider(
    names_from  = Cluster,
    values_from = Freq,
    values_fill = 0
  )

wide_df <- wide_df %>%
  mutate(
    ratio_rBEC = ifelse(BEC + rBEC > 0, rBEC / (BEC + rBEC), NA_real_),
    ratio_rFRC = ifelse(FRC + rFRC + FDC > 0, rFRC / (FRC + rFRC + FDC), NA_real_)
  )

wide_df$ratio_rBEC <- wide_df$ratio_rBEC *100
wide_df$ratio_rFRC <- wide_df$ratio_rFRC *100

wide_df$Group <- factor(wide_df$Group, levels = c("rLN", "FL", "DLBCL"))
my_comparisons <- list( c("rLN", "FL"), c("FL", "DLBCL"), c("rLN", "DLBCL"))

## Generate plots -----------------------------------------------------------
p1 <- ggplot(wide_df, aes(Group, ratio_rBEC)) +
  geom_boxplot(aes(fill = Group), width = 0.5, outlier.shape = NA) +
  ggbeeswarm::geom_beeswarm(
    method = "compactswarm",     # tighter, more even spreading
    priority = "density",        # order points by local density -> less stacking
    corral = "wrap",             # keep all points inside the width
    cex = 3, 
    corral.width = 0.5,          # match your boxplot width
    dodge.width = 0.5,           # align with grouped boxes if any
    size = 2, shape = 1, stroke = 0.75
  ) +
  scale_fill_manual(values = c("#8D60B5", "#8BCDE0", "#67B86E")) +
  scale_y_continuous(limits = c(0,110), breaks = c(0,25,50,75,100))+
  ylab("% rBEC") +
  guides(fill = guide_legend(title = "Entity")) +
  theme_bw() +
  theme(text = element_text(size=11), 
        legend.position = "none",
        axis.text.y = element_text(size = 12, color = "black"),
        axis.text.x = element_text(size = 12, color = "black"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 12, color = "black"),
        #legend.title = element_blank(),
        #legend.text = element_text(size = 12, color = "black"),
        plot.title = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        aspect.ratio = 1.5)+ 
  stat_compare_means(method = "wilcox.test", label = "p.format", label.y = c(85,95,105),
                     comparisons = my_comparisons,
                     hide.ns = FALSE)

p2 <- ggplot(wide_df, aes(Group, ratio_rFRC)) +
  geom_boxplot(aes(fill = Group), width = 0.5, outlier.shape = NA) +
  ggbeeswarm::geom_beeswarm(
    method = "compactswarm",     # tighter, more even spreading
    priority = "density",        # order points by local density -> less stacking
    corral = "wrap",             # keep all points inside the width
    cex = 3, 
    corral.width = 0.5,          # match your boxplot width
    dodge.width = 0.5,           # align with grouped boxes if any
    size = 2, shape = 1, stroke = 0.75
  ) +
  scale_fill_manual(values = c("#8D60B5", "#8BCDE0", "#67B86E")) +
  scale_y_continuous(limits = c(0,115), breaks = c(0,25,50,75,100))+
  ylab("% rFRC") +
  guides(fill = guide_legend(title = "Entity")) +
  theme_bw() +
  theme(text = element_text(size=11), 
        legend.position = "none",
        axis.text.y = element_text(size = 12, color = "black"),
        axis.text.x = element_text(size = 12, color = "black"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 12, color = "black"),
        #legend.title = element_blank(),
        #legend.text = element_text(size = 12, color = "black"),
        plot.title = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        aspect.ratio = 1.5)+ 
  stat_compare_means(method = "wilcox.test", label = "p.format", label.y = c(85,95,105),
                     comparisons = my_comparisons,
                     hide.ns = FALSE)

combined_plot <- p2 / p1

print(combined_plot)
```

### B
```{r Panel C, fig.height=3, fig.width=6}
#Differential expression analysis
#Perform DE analysis between DLBCL and rLN
dlbcl_rln_obj <- subset(scRNA, subset = group %in% c("rLN", "DLBCL"))
DefaultAssay(dlbcl_rln_obj) <- "RNA"
Idents(dlbcl_rln_obj) <- "Clusters_binned"

DE_genes <- list()
DE_genes_entrez <- list()

#Ran for all populations since these data are eventually used for supplemental figures
for (j in c("FRC", "BEC", "LEC")){
  jj <- str_replace(j, "/", "_")
  tmp_obj <- subset(dlbcl_rln_obj, subset = Clusters_binned == j)
  Idents(tmp_obj) <- "group"
  DE_all <- FindMarkers(tmp_obj, 
                        ident.2 = "rLN", 
                        ident.1 = "DLBCL", 
                        pseudocount.use = 0.01, 
                      test.use = "MAST", 
                      only.pos = F, 
                      logfc.threshold = 0,
                      assay = "RNA")
  write.table(DE_all, file=paste0(out_dir, jj,"_DEgenes_MAST_dlbclVSrln.tsv"), 
              quote=FALSE, 
              sep="\t", 
              row.names = T)
}
#Specifically code for FRC
de_frc_rln_dlbcl <- read.table(paste0(out_dir, "FRC_DEgenes_MAST_dlbclVSrln.tsv"))
mt_genes <- grep("^MT-",rownames(de_frc_rln_dlbcl))
rp_genes <- c(grep("^RPS",rownames(de_frc_rln_dlbcl)), 
              grep("^RPL",rownames(de_frc_rln_dlbcl)))
de_frc_rln_dlbcl <- de_frc_rln_dlbcl[-c(mt_genes, rp_genes), ]
de_frc_rln_dlbcl$gene <- rownames(de_frc_rln_dlbcl)
colnames(de_frc_rln_dlbcl) <- c("p_val", "log2FoldChange", "pct.1", "pct.2", "pvalue", "gene")
```

```{r Panel C, fig.width=3, fig.height=3}
de_frc_rln_dlbcl$diffexpressed <- "NO"
# if log2Foldchange > 0.6 and pvalue < 0.05, set as "UP" 
de_frc_rln_dlbcl$diffexpressed[de_frc_rln_dlbcl$log2FoldChange > 0.5 & de_frc_rln_dlbcl$pvalue < 10^-3] <- "UP"
# if log2Foldchange < -0.6 and pvalue < 0.05, set as "DOWN"
de_frc_rln_dlbcl$diffexpressed[de_frc_rln_dlbcl$log2FoldChange < -0.5 & de_frc_rln_dlbcl$pvalue < 10^-3] <- "DOWN"

de_frc_rln_dlbcl$diffexpressed[de_frc_rln_dlbcl$gene %in% c("CXCL13", "CXCL12", "CCL21", "CCL19")] <- "DOWN"
de_frc_rln_dlbcl$diffexpressed[de_frc_rln_dlbcl$gene %in% c("CXCL9", "CXCL10", "CXCL11")] <- "UP"

`%!in%` = Negate(`%in%`)
de_frc_rln_dlbcl$delabel <- NA
de_frc_rln_dlbcl$delabel[de_frc_rln_dlbcl$diffexpressed %!in% c("NO", "DOWN")] <- de_frc_rln_dlbcl$gene[de_frc_rln_dlbcl$diffexpressed %!in% c("NO", "DOWN")]

de_frc_rln_dlbcl$delabel <- ifelse(de_frc_rln_dlbcl$gene == "CXCL13"
                                  | de_frc_rln_dlbcl$gene == "CXCL12"
                                  | de_frc_rln_dlbcl$gene == "CCL21"
                                  | de_frc_rln_dlbcl$gene == "CCL19"
                                  | de_frc_rln_dlbcl$gene == "CXCL9"
                                  | de_frc_rln_dlbcl$gene == "CXCL10"
                                  | de_frc_rln_dlbcl$gene == "CXCL11",T, F)

ggplot(data=de_frc_rln_dlbcl, aes(x=log2FoldChange, y=-log10(pvalue), col=diffexpressed)) + 
  geom_point(alpha = 0.6, size = 1.5) + 
  ggrepel::geom_text_repel(data = de_frc_rln_dlbcl, 
                           aes(label = ifelse(delabel == TRUE, 
                                     as.character(gene),"")), 
                           box.padding = 0.5, 
                           max.overlaps = Inf,
                           min.segment.length = 0,
                           hjust= 0.3,
                           colour = "black")+
  #geom_label_repel(aes(label = gene),
  #                 box.padding   = 1)+
  geom_vline(xintercept = c(-0.5, 0.5), linetype = 2)+
  labs(x ="log2(FC)", y = "-log10(p.adj)")+
  theme_bw() +
  theme(text = element_text(size=11), 
        axis.text.y = element_text(size = 11, color = "black"),
        axis.text.x = element_text(size = 11, color = "black"),
        legend.title = element_blank(),
        plot.title = element_text("tipBEC vs. other BEC"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        aspect.ratio = 1,
        legend.position="none") +
  scale_color_manual(values=c("#8D60B5", "#D3D3D3", "#67B86E"))
```

## C
```{r Panel C}
loeffler <- readRDS("") 
#Data from Loeffler-Wirth, H. et al. A modular transcriptome map of mature B cell lymphomas. Genome Med. 11, 27 (2019).
#Full object available upon request

diagnosis_colors <- c("Tonsil" = "#8D60B5", "FL 1/2/3A" = "#8BCDE0", "FL 3B" = "#089c9c", "DLBCL" = "#67B86E")

# Define chemokine scores
metadata <- as.data.frame(loeffler@meta.data)
genes <- rownames(loeffler@assays$RNA)[rownames(loeffler@assays$RNA) %in% c("CXCL12", "CCL21", "CCL19", "CXCL13", "CXCL9", "CXCL10", "CXCL11")]
metadata <-  cbind(metadata, FetchData(loeffler, genes))

metadata$ifn_mean <- rowMeans(metadata[,c("CXCL9", "CXCL10", "CXCL11")], na.rm=TRUE)
metadata$homeo_mean <- rowMeans(metadata[,c("CXCL12", "CCL21", "CCL19", "CXCL13")], na.rm=TRUE)

metadata$OS_status <- as.numeric(as.character(metadata$OS_status))

# Tidy disease annotation
metadata <- mutate(metadata, Type_binned2 = case_when(Type_binned %in% c("DLBCL ABC", "DLBCL GCB", "DLBCL unclassified")~"DLBCL",
                                                           Type_binned %in% c("FL 3b")~"FL 3B",
                                                           Type_binned == "Tonsil"~"Tonsil",
                                                           Type_binned == "FL 1/2/3a"~"FL 1/2/3A"))

metadata$Type_binned2 <- factor(metadata$Type_binned2, levels = c("Tonsil", "FL 1/2/3A", "FL 3B", "DLBCL"))
comp <- list(c("Tonsil", "FL 1/2/3A"), c("FL 3B", "FL 1/2/3A"),  c("DLBCL", "FL 1/2/3A"), c("DLBCL", "FL 3B"), c("Tonsil", "DLBCL"))


p1 <- ggplot(metadata, aes(Type_binned2, homeo_mean)) + 
    geom_boxplot(aes(fill=Type_binned2), outlier.size=0.5, width = 0.5) + 
    geom_quasirandom(method = "pseudorandom", size = 1, alpha = 0.3)+
    labs(y = "Mean gene expression", 
         x = "") + 
    scale_fill_manual(values=diagnosis_colors) +
    theme(axis.title.x = element_text(size = 12), 
          axis.title.y = element_text(size = 12)) +
    scale_y_continuous(breaks=c(0,1,2,3,4), limits = c(0,5.5))+
    ggtitle("Homeostatic chemokines") +
    theme_bw()+
    theme(
      plot.title = element_text(size = 13, face = "bold"),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.background = element_rect(color = "black"), 
      legend.position = "none",
      axis.text.y = element_text(size = 13, color = "black"),
      axis.text.x = element_text(size = 13, color = "black"),
      axis.title.x = element_text(size = 13, color = "black"),
      aspect.ratio = 1,
      axis.title.y = element_text(size = 13, color = "black"))+
    stat_compare_means(method = "t.test", label = "p.signif", comparisons = comp, 
                       label.y = c(3.8, 3.8, 3.6, 4.2, 4.6), hide.ns = F)


p2 <- ggplot(metadata, aes(Type_binned2, ifn_mean)) + 
    geom_boxplot(aes(fill=Type_binned2), outlier.size=0.5, width = 0.5) + 
    geom_quasirandom(method = "pseudorandom", size = 1, alpha = 0.3)+
    labs(y = "Mean gene expression", 
         x = "") + 
    scale_fill_manual(values=diagnosis_colors) +
    theme(axis.title.x = element_text(size = 12), 
          axis.title.y = element_text(size = 12)) +
    scale_y_continuous(breaks=c(0,1,2,3,4), limits = c(0,5.5))+
    ggtitle("Inflammatory chemokines") +
    theme_bw()+
    theme(
      plot.title = element_text(size = 13, face = "bold"),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.background = element_rect(color = "black"), 
      legend.position = "none",
      axis.text.y = element_text(size = 13, color = "black"),
      axis.text.x = element_text(size = 13, color = "black"),
      axis.title.x = element_text(size = 13, color = "black"),
      aspect.ratio = 1,
      axis.title.y = element_text(size = 13, color = "black"))+
    stat_compare_means(method = "t.test", label = "p.signif", comparisons = comp, 
                       label.y = c(3.8, 3.8, 3.6, 4.2, 4.6), hide.ns = F)

p1/p2
```

## H and I
```{r Panels H, I }
#For Xenium related plots and analysis, also refer to https://github.com/ZauggGroup/LNarch/edit/main/Preprocessing/20250304_xenium.Rmd

# Load data
xenium_seurat <- readRDS("") #Available in zenodo as "xenium_merged_seurat.rds"

# Left panels (cell types)

fl <- subset(merged_obj, type %in% c("FL"))
metadata_fl <- as.data.frame(fl@meta.data)
DefaultAssay(fl) <- "RNA"
genes <- c("CXCL13", "CXCL12", "CCL21", "CCL19", "CXCL9", "CXCL10", "CXCL11", "CXCR3", "IFNG")
fl_df <-  cbind(metadata_fl, FetchData(fl, genes))

panel_h_left <- ggplot(fl_df, aes(x = nucleus_centroid_x, y = nucleus_centroid_y, color = cell_type_binned)) +
  # Plot the background points with #f0f0f0
  geom_point(data = subset(fl_df, cell_type_binned %in% c("BEC", "DLBCL-B", "Myeloid", "T-TOX", "T-REG", "T-FH", "T-H")),
             aes(alpha = 1), size = 0.75, color = "#f0f0f0", shape = 16) +
  
    # Plot the "FL-B" cells with alpha = 0
  geom_point(data = subset(fl_df, cell_type_binned == "FL-B"),
             aes(alpha = 0), size = 0.75, color = "#0EBCE8", shape = 16) +
  
  # Plot all other colored points with alpha = 1
  geom_point(data = subset(fl_df, cell_type_binned %in% c("FRC", "rFRC", "FDC")),
             aes(alpha = 1), size = 0.75, shape = 16) +
  scale_color_manual(values = c(
    "BEC" = "#f0f0f0",
    "Myeloid" = "#f0f0f0",
    "T-TOX" = "#f0f0f0",
    "T-REG"= "#f0f0f0", 
    "T-FH"= "#f0f0f0", 
    "T-H" = "#f0f0f0",
    "DLBCL-B" = "#f0f0f0",
    "FL-B" = "#0EBCE8",
    "FRC" = "#BEDEA6",
    "rFRC" = "#2AC200",
    "FDC" = "#088B1A"))+
  theme_bw() +
  theme(legend.position = "none",
        aspect.ratio = 1,
        panel.grid.minor = element_blank(),
        panel.grid.major = element_blank(),
        panel.background = element_rect(color = "black"))


dlbcl <- subset(merged_obj, type %in% c("DLBCL"))
metadata_dlbcl <- as.data.frame(dlbcl@meta.data)
DefaultAssay(dlbcl) <- "RNA"
dlbcl_df <-  cbind(metadata_dlbcl, FetchData(dlbcl, genes))

panel_i_left <- ggplot(dlbcl_df, aes(x = nucleus_centroid_x, y = nucleus_centroid_y, color = cell_type_binned)) +
  # Plot the background points with #f0f0f0
  geom_point(data = subset(dlbcl_df, cell_type_binned %in% c("BEC", "FL-B", "Myeloid", "T-TOX", "T-REG", "T-FH", "T-H")),
             aes(alpha = 1), size = 0.75, color = "#f0f0f0", shape = 16) +
  
    # Plot the "DLBCL-B" cells with alpha = 0
  geom_point(data = subset(dlbcl_df, cell_type_binned == "DLBCL-B"),
             aes(alpha = 0), size = 0.75, color = "#83CAE8", shape = 16) +
  
  # Plot all other colored points with alpha = 1
  geom_point(data = subset(dlbcl_df, cell_type_binned %in% c("FRC", "rFRC", "FDC")),
             aes(alpha = 1), size = 0.75, shape = 16) +
  scale_color_manual(values = c(
    "BEC" = "#f0f0f0",
    "Myeloid" = "#f0f0f0",
    "T-TOX" = "#f0f0f0",
    "T-REG"= "#f0f0f0", 
    "T-FH"= "#f0f0f0", 
    "T-H" = "#f0f0f0",
    "FL-B" = "#f0f0f0",
    "DLBCL-B" = "#83CAE8",
    "FRC" = "#BEDEA6",
    "rFRC" = "#2AC200",
    "FDC" = "#088B1A"))+
  theme_bw() +
  theme(legend.position = "none",
        panel.grid.minor = element_blank(),
        panel.grid.major = element_blank(),
        panel.background = element_rect(color = "black"),
        aspect.ratio = 1)

# Right panels (CXCL13 - CXCR5

lr_fl <- read.csv("path/to/.csv") #available in source data for Figure 2
lr_dlbcl <- read.csv("path/to/.csv") #available in source data for Figure 2

lr_fl <- lr_fl %>%
  mutate(order = ifelse(ligand_receptor_assoc %in% c(1,2,3,4,5), 1, 0))

lr_dlbcl <- lr_dlbcl %>%
  mutate(order = ifelse(ligand_receptor_assoc %in% c(1,2,3,4,5), 1, 0))

panel_h_right <- ggplot(lr_fl, aes(x = nucleus_centroid_x.x, y = nucleus_centroid_y.x, color = ligand_receptor_assoc)) +
  geom_point(data = subset(lr_fl, order == 0),
             size = 0.5, color = "#f9f9f9", alpha = 0.1) +
  geom_point(data = subset(lr_fl, order == 1),
             size = 0.5, aes(color = ligand_receptor_assoc), alpha = 1) +
  #geom_point(size = 0.5, alpha = 0.7) +
  scale_color_gradientn(colors = c("#f9f9f9", "#AD2E24")) +
  theme_bw() +
  theme(legend.position = "none",
        panel.grid.minor = element_blank(),
        panel.grid.major = element_blank(),
        panel.background = element_rect(color = "black"),
        aspect.ratio = 1)

panel_i_right <- ggplot(lr_dlbcl, aes(x = nucleus_centroid_x.x, y = nucleus_centroid_y.x, color = ligand_receptor_assoc)) +
  geom_point(data = subset(lr_dlbcl, order == 0),
             size = 0.5, color = "#f9f9f9", alpha = 1) +
  geom_point(data = subset(lr_dlbcl, order == 1),
             size = 0.5, aes(color = ligand_receptor_assoc), alpha = 1) +
  scale_color_gradientn(colors = c("#f9f9f9", "#AD2E24")) +
  theme_bw() +
  theme(legend.position = "none",
        panel.grid.minor = element_blank(),
        panel.grid.major = element_blank(),
        panel.background = element_rect(color = "black"),
        aspect.ratio = 1)


panel_h_left + panel_h_right
panel_i_left + panel_i_right
```

```{r Panel K}
#For TCR seq related analysis, see https://github.com/ZauggGroup/LNarch/edit/main/Preprocessing/scTCR_reanalysis_revisions.Rmd
## Object also used for TCR clonal expansion analysis in Figure 3
## Dataset from Roider et al. Nat Cell Biol 2024
## 5' TCR seq single cell object projected to 3' scRNA-seq dataset similarly to the original publication
## Analysis was focused on rLN, FL and DLBCL
```
